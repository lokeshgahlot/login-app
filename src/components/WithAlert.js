import React from "react";
import PropTypes from "prop-types";
import {connect} from "react-redux";
import {Alert} from "react-bootstrap";
import {withRouter} from "react-router-dom";
import {clear} from "../actions/alert.actions";

export class AppAlert extends React.Component {
  constructor(props) {
    super(props);
    const {history} = props;
    this.historyListener = this.historyListener.bind(this);
    history.listen(this.historyListener);
  }

  historyListener(location, action) {
    const {clearAlert} = this.props;
    clearAlert();
  }

  componentWillUnmount() {
    const {history} = this.props;
    history.unlisten(this.historyListener);
  }

  render() {
    const {bsStyle, message, children, dispatch, staticContext, clearAlert, ...rest} = this.props;
    return (
      <React.Fragment>
        {
          message.length > 0 ? <Alert bsStyle={bsStyle} {...rest}>{message}</Alert> : null
        }
        {children}
      </React.Fragment>
    );
  }
}

AppAlert.propTypes = {
  bsStyle: PropTypes.string,
  message: PropTypes.string
};

AppAlert.defaultProps = {
  bsStyle: "danger",
  message: ""
};


const mapSTP = (state) => ({
  bsStyle: state.get("alert").get("type"),
  message: state.get("alert").get("message")
});

//Added with withRouter because of the following reason 
// https://github.com/ReactTraining/react-router/blob/master/packages/react-router/docs/guides/redux.md#blocked-updates
export default withRouter(
  connect(mapSTP, {clearAlert: clear})
  (AppAlert)
);



