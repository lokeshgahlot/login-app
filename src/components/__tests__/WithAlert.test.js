import React from "react";
import {shallow} from "enzyme"
import {fromJS} from "immutable";
import {AppAlert, mapSTP} from "../WithAlert";

describe("<App />", () => {
  let mockHistory;
  beforeEach(() => {
    mockHistory = {
      listen: jest.fn()
    }
  });

  afterEach(() => {
    mockHistory.listen.mockClear();
    mockHistory = undefined;
  });

  it ("renders without alert", () => {
    const wrapper = shallow(
      <AppAlert history={mockHistory}>
        <span>hello</span>
      </AppAlert>
    );
    expect(wrapper).toMatchSnapshot();
  });

  it ("renders with alert", () => {
    const wrapper = shallow(
      <AppAlert history={mockHistory} message="Test alert!!" bsStyle="info">
        <span>hello</span>
      </AppAlert>
    );
    expect(wrapper).toMatchSnapshot();
  });

  it ("adds history listener", () => {
    const wrapper = shallow(
      <AppAlert history={mockHistory}>
        <span>hello</span>
      </AppAlert>
    );
    const instance = wrapper.instance();
    expect(mockHistory.listen.mock.calls.length).toBe(1);
    // listener function
    expect(mockHistory.listen.mock.calls[0][0]).toEqual(instance.historyListener);
  });

  it ("AppAlert.historyListener() calls clearAlert()", () => {
    const fn = jest.fn();
    const wrapper = shallow(
      <AppAlert history={mockHistory} clearAlert={fn}>
        <span>hello</span>
      </AppAlert>
    );
    const instance = wrapper.instance();
    instance.historyListener();
    expect(fn.mock.calls.length).toBe(1);
  });

  it ("mapSTP()", () => {
    const state = fromJS({
      alert: {
        type: "info",
        message: "Test Alert!!"
      }
    });
    expect(mapSTP(state)).toEqual({
      bsStyle: "info",
      message: "Test Alert!!"
    });
  });

});