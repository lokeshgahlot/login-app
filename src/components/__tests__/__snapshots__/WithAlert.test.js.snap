// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> renders with alert 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <AppAlert
    bsStyle="info"
    history={
      Object {
        "listen": [MockFunction] {
          "calls": Array [
            Array [
              [Function],
            ],
          ],
        },
      }
    }
    message="Test alert!!"
  >
    <span>
      hello
    </span>
  </AppAlert>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <Alert
          bsClass="alert"
          bsStyle="info"
          closeLabel="Close alert"
          history={
            Object {
              "listen": [MockFunction] {
                "calls": Array [
                  Array [
                    [Function],
                  ],
                ],
              },
            }
          }
        >
          Test alert!!
        </Alert>,
        <span>
          hello
        </span>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "bsClass": "alert",
          "bsStyle": "info",
          "children": "Test alert!!",
          "closeLabel": "Close alert",
          "history": Object {
            "listen": [MockFunction] {
              "calls": Array [
                Array [
                  [Function],
                ],
              ],
            },
          },
        },
        "ref": null,
        "rendered": "Test alert!!",
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "hello",
        },
        "ref": null,
        "rendered": "hello",
        "type": "span",
      },
    ],
    "type": Symbol(react.fragment),
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <Alert
            bsClass="alert"
            bsStyle="info"
            closeLabel="Close alert"
            history={
              Object {
                "listen": [MockFunction] {
                  "calls": Array [
                    Array [
                      [Function],
                    ],
                  ],
                },
              }
            }
          >
            Test alert!!
          </Alert>,
          <span>
            hello
          </span>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "bsClass": "alert",
            "bsStyle": "info",
            "children": "Test alert!!",
            "closeLabel": "Close alert",
            "history": Object {
              "listen": [MockFunction] {
                "calls": Array [
                  Array [
                    [Function],
                  ],
                ],
              },
            },
          },
          "ref": null,
          "rendered": "Test alert!!",
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "hello",
          },
          "ref": null,
          "rendered": "hello",
          "type": "span",
        },
      ],
      "type": Symbol(react.fragment),
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`<App /> renders without alert 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <AppAlert
    bsStyle="danger"
    history={
      Object {
        "listen": [MockFunction] {
          "calls": Array [
            Array [
              [Function],
            ],
          ],
        },
      }
    }
    message=""
  >
    <span>
      hello
    </span>
  </AppAlert>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        null,
        <span>
          hello
        </span>,
      ],
    },
    "ref": null,
    "rendered": Array [
      null,
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "hello",
        },
        "ref": null,
        "rendered": "hello",
        "type": "span",
      },
    ],
    "type": Symbol(react.fragment),
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          null,
          <span>
            hello
          </span>,
        ],
      },
      "ref": null,
      "rendered": Array [
        null,
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "hello",
          },
          "ref": null,
          "rendered": "hello",
          "type": "span",
        },
      ],
      "type": Symbol(react.fragment),
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
